### PR TITLE
#9208: Functional SqueezeBERT model Demo

### PR DESCRIPTION
**Pull request by kkeerthana0573**
### Ticket
- [Link to Github Issue](https://github.com/tenstorrent/tt-metal/issues/9208).

### Problem description
- Functional SqueezeBERT model Demo.

### What's changed
- TTNN support for SqueezeBERT model.
- Added demo for SqueezeBertForQuestionAnswering task.

### Checklist
- [x] Post commit CI passes 
- [ ] Blackhole Post commit (if applicable)
- [x] Model regression CI testing passes (if applicable)
- [x] Device performance regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

Links to the passing CIs (09/12/2024):
 - [All post-commit tests](https://github.com/tenstorrent/tt-metal/actions/runs/12228869079)
 -  [(Single-card) Demo tests](https://github.com/tenstorrent/tt-metal/actions/runs/12228870281) 
-  (Single-card) Device perf regressions : [GS](https://github.com/tenstorrent/tt-metal/actions/runs/12228871915/job/34107907576#step:9:1054), [WH](https://github.com/tenstorrent/tt-metal/actions/runs/12228871915/job/34107907739#step:9:1050)
- (Single-card) Model perf tests : [GS](https://github.com/tenstorrent/tt-metal/actions/runs/12228873639/job/34107873770#step:10:1541), [WH](https://github.com/tenstorrent/tt-metal/actions/runs/12228873639/job/34107874180#step:10:1263)
- [Nightly fast dispatch tests](https://github.com/tenstorrent/tt-metal/actions/runs/12228875825)

### CODE REVIEW

**Review by eyonland — CHANGES_REQUESTED** [review]
I see the e2e performance tests but not seeing a test marked with @pytest.mark.models_device_performance_bare_metal.

Note that we document this here: https://tenstorrent.github.io/tt-metal/latest/ttnn/ttnn/demos.html
See for example:
tests/ttnn/integration_tests/resnet/test_performance.py

**Review by TT-BrianLiu — APPROVED** [review]

**Comment by kkeerthana0573** [comment]
@eyonland, 
Will add the test and update PR.
Thank you.

File: models/experimental/functional_squeezebert/README.md
Comment (tt-rkim) [inline models/experimental/functional_squeezebert/README.md:4]: Please put clear indications on what TT platforms (GS, WH N150, WH N300) these demos run on.

Refer to other demo READMEs for what I'm talking about, particularly in `models/demos/wormhole`.

File: tests/ttnn/integration_tests/squeezebert/test_ttnn_squeezebert.py
Comment (tt-rkim) [inline tests/ttnn/integration_tests/squeezebert/test_ttnn_squeezebert.py:20]: Why is there a `test_ttnn_squeezebert.py` in here and in `models/experimental`? Why do we need so many?

File: tests/ttnn/integration_tests/squeezebert/test_performance.py
Comment (tt-rkim) [inline tests/ttnn/integration_tests/squeezebert/test_performance.py:15]: `tests/ttnn/integration_tests` is a poor place for a performance test. there are some that are already in here.

That is not recommended. Please move it to a different folder within `models/experiemental/...`.

Comment (tt-rkim) [reply tests/ttnn/integration_tests/squeezebert/test_performance.py:15]: Any action on this?

File: tests/ttnn/integration_tests/squeezebert/test_ttnn_squeezebert.py
Comment (tt-rkim) [reply tests/ttnn/integration_tests/squeezebert/test_ttnn_squeezebert.py:20]: Any action on this?

Comment (kkeerthana0573) [reply tests/ttnn/integration_tests/squeezebert/test_ttnn_squeezebert.py:20]: The test is removed from `models/experimental/` and is now only in `tests/ttnn/integration_tests/`.
Corresponding commit - [link](https://github.com/tenstorrent/tt-metal/compare/main...keerthana/functional_squeezebert_). 
Thank you.

File: tests/ttnn/integration_tests/squeezebert/test_performance.py
Comment (kkeerthana0573) [reply tests/ttnn/integration_tests/squeezebert/test_performance.py]: test_performance is moved to `models/experimental/functional_squeezebert/tests/`. 
Commit link - [link](https://github.com/tenstorrent/tt-metal/compare/main...keerthana/functional_squeezebert_).
Thank you.

**Comment by kkeerthana0573** [comment]
@eyonland, 
The test for device performance is added. 
Test link - [test_perf_device_squeezebert](https://github.com/tenstorrent/tt-metal/compare/main...keerthana/functional_squeezebert_#diff-ee52992802fc4ae5ca8f49fa1c765845e5cc685f3eb200929b59fc25f02bd6db)
Thank you.

File: models/experimental/functional_squeezebert/tests/test_performance.py
Comment (tt-rkim) [inline models/experimental/functional_squeezebert/tests/test_performance.py:61]: Are you trying to run e2e performance test? It's not specified in any scripts

Comment (kkeerthana0573) [reply models/experimental/functional_squeezebert/tests/test_performance.py:61]: Updated `run_performance.sh` with e2e performance test along with device perf test.

**Comment by tt-rkim** [comment]
Please post passing fast dispatch nightly, model perf, and device perf pipelines. Because you only recently added the commands, you must run them again.

File: tests/scripts/run_performance.sh
Comment (tt-rkim) [inline tests/scripts/run_performance.sh:33]: Too many newlines

File: tests/scripts/run_performance.sh
Comment (tt-rkim) [inline tests/scripts/run_performance.sh:101]: Why extra newline?

**Comment by tt-rkim** [comment]
Where are the passing pipelines?

Single card demo
Single card nightly fast dispatch
Single card model perf
Single card device perf

**Comment by kkeerthana0573** [comment]
@tt-rkim, 
Passing CI links: 

- [All post-commit tests](https://github.com/tenstorrent/tt-metal/actions/runs/11211478449)
- [(Single-card) Demo tests](https://github.com/tenstorrent/tt-metal/actions/runs/11211484821)
- [Nightly fast dispatch tests](https://github.com/tenstorrent/tt-metal/actions/runs/11211511180)
- [(Single-card) Device perf regressions](https://github.com/tenstorrent/tt-metal/actions/runs/11211496525/job/31161151615#step:9:818)
- [(Single-card) Model perf tests](https://github.com/tenstorrent/tt-metal/actions/runs/11211499942)

File: tests/scripts/run_performance.sh
Comment (kkeerthana0573) [reply tests/scripts/run_performance.sh]: Removed extra lines.

File: tests/scripts/run_performance.sh
Comment (kkeerthana0573) [reply tests/scripts/run_performance.sh:33]: Removed extra lines.

**Comment by kkeerthana0573** [comment]
Passing CI links:

[All post-commit tests](https://github.com/tenstorrent/tt-metal/actions/runs/11549521498)
[(Single-card) Demo tests](https://github.com/tenstorrent/tt-metal/actions/runs/11549523350)
[Nightly fast dispatch tests](https://github.com/tenstorrent/tt-metal/actions/runs/11549531267)
[(Single-card) Device perf regressions](https://github.com/tenstorrent/tt-metal/actions/runs/11549525677/job/32142829567#step:9:940)
[(Single-card) Model perf tests](https://github.com/tenstorrent/tt-metal/actions/runs/11549527898)

**Comment by tt-rkim** [comment]
Did those device perf checks actually happen?

**Comment by kkeerthana0573** [comment]
@tt-rkim, 
Yes, the device performance checks did happen for the model. Here is the link for the model passing in CI: [link](https://github.com/tenstorrent/tt-metal/actions/runs/11909699702/job/33188080549#step:9:722)

**Comment by tt-rkim** [comment]
`2024-11-19 11:56:14.853 | ERROR    | models.perf.device_perf_utils:check_device_perf:67 - AVG DEVICE KERNEL SAMPLES/S 186.23090621076346 is outside of expected range (198.559, 210.84099999999998)`

**Review by eyonland — APPROVED** [review]
Please make sure all your tests pass before merging.

**Comment by kkeerthana0573** [comment]
@tt-rkim, 
I've updated deviceperf expected range. Here is the passing Device Perf CI [link](https://github.com/tenstorrent/tt-metal/actions/runs/11928473947/job/33245906399#step:9:765). 
Thank you.

**Comment by kkeerthana0573** [comment]
> Please make sure all your tests pass before merging.


@eyonland, 
Sure, We'll merge only after the tests pass.

**Comment by tt-rkim** [comment]
What happened here? https://github.com/tenstorrent/tt-metal/actions/runs/11928473947/job/33245906399#step:9:765

**Comment by tt-rkim** [comment]
I don't see the post.

**Comment by kkeerthana0573** [comment]
@tt-rkim,
The CI was passing for squeezebert [here](https://github.com/tenstorrent/tt-metal/actions/runs/11928473947/job/33245906399#step:9:765). Do you suspect any issue here?

**Comment by kkeerthana0573** [comment]
Squeezebert's device performance test on n300 passed [here](https://github.com/tenstorrent/tt-metal/actions/runs/11928473947/job/33245906399#step:9:780).

**Comment by tt-rkim** [comment]
Why are you not addressing the error I'm mentioning?

This line in your PR: https://github.com/tenstorrent/tt-metal/pull/9371/files#diff-17b43eb99e1634341d15c11cff11877200bdc21c12d4a70e05ad8630ac2117afR128

is causing a file not found error

**Comment by kkeerthana0573** [comment]
@tt-rkim, PR is updated. 


Passing CI links:
[All post-commit tests](https://github.com/tenstorrent/tt-metal/actions/runs/11984636135)
[(Single-card) Demo tests](https://github.com/tenstorrent/tt-metal/actions/runs/11984637587)
[Nightly fast dispatch tests](https://github.com/tenstorrent/tt-metal/actions/runs/11984640014)
[(Single-card) Device perf regressions](https://github.com/tenstorrent/tt-metal/actions/runs/11985446754)
[(Single-card) Model perf tests](https://github.com/tenstorrent/tt-metal/actions/runs/11984638802)

**Comment by tt-rkim** [comment]
Please update the threshold:

```
models.perf.device_perf_utils:check_device_perf:67 - AVG DEVICE KERNEL SAMPLES/S 283.90111071341295 is outside of expected range (180.61399999999998, 191.786)
```

for your model

I will approve to unblock

**Review by tt-rkim — APPROVED** [review]

File: models/demos/squeezebert/demo/demo.py
Comment (uaydonat) [inline models/demos/squeezebert/demo/demo.py:257]: it would be safer if the test fails when eval_score is not close to cpu_eval_score by some small threshold.

File: models/demos/squeezebert/demo/demo.py
Comment (uaydonat) [inline models/demos/squeezebert/demo/demo.py:275]: default batch size should be doubled when running on n300

File: models/demos/squeezebert/demo/demo.py
Comment (uaydonat) [inline models/demos/squeezebert/demo/demo.py:297]: same. batch_size should be double on n300 than on n150

File: models/demos/squeezebert/tests/test_perf_device_squeezebert.py
Comment (uaydonat) [inline models/demos/squeezebert/tests/test_perf_device_squeezebert.py:21]: what;'s the current device perf we are getting? how close is it to expected_perf?

File: models/demos/squeezebert/tests/test_performance.py
Comment (uaydonat) [inline models/demos/squeezebert/tests/test_performance.py:50]: what's the current device time? how close is it to expected times?

**Review by uaydonat — CHANGES_REQUESTED** [review]

File: models/demos/squeezebert/demo/demo.py
Comment (kkeerthana0573) [reply models/demos/squeezebert/demo/demo.py:275]: This is a single-card demo implementation, and we expect the same batch size for both N150 and for single device on N300.

File: models/demos/squeezebert/demo/demo.py
Comment (kkeerthana0573) [reply models/demos/squeezebert/demo/demo.py:297]: This is a single-card demo implementation, and we expect the same batch size for both N150 and for single device on N300. We have a data parallel implementation where the batch size is doubled for N300 - [link](https://github.com/tenstorrent/tt-metal/pull/13418).

File: models/demos/squeezebert/tests/test_perf_device_squeezebert.py
Comment (kkeerthana0573) [reply models/demos/squeezebert/tests/test_perf_device_squeezebert.py:21]: The current expected_perf in the test file have been updated based on the device perf observed during CI runs.

File: models/demos/squeezebert/demo/demo.py
Comment (kkeerthana0573) [reply models/demos/squeezebert/demo/demo.py:257]: I've updated the demo to assert eval_scores against cpu_eval_scores within a 3% margin.

**Review by uaydonat — APPROVED** [review]

**Comment by kkeerthana0573** [comment]
Updated description with passing CI links and merged.

**Pull request merged**
